### DISCUSSION TITLE
Content length is zero for pages created with AutoConnect

### DISCUSSION THREAD

**Discussion by faanskit** [Hieromon/AutoConnect#556] [Q&A] [answered]
I've noted the following warning for each request to AutoConnect portal pages created by me.

```
[ 79445][W][WebServer.cpp:436] send(): content length is zero
[ 82495][W][WebServer.cpp:436] send(): content length is zero
[ 85596][W][WebServer.cpp:436] send(): content length is zero
[354996][W][WebServer.cpp:436] send(): content length is zero
[358046][W][WebServer.cpp:436] send(): content length is zero
```
A simple example page:
```c++
// Building and creating the Configuration resul page
AutoConnectAux* MyPortal::configResultPage(){
        AutoConnectElement* myText = new AutoConnectText("area_results", "","","", AC_Tag_None);
        AutoConnectElement* myDone = new AutoConnectSubmit("done", "Klar", HOME_PAGE);
        return new AutoConnectAux(CONFIG_DONE_PAGE, CONFIG_PAGE_HEADER, false, {*myText, *myDone});;
}
```

I do use a callback:
```c++
AutoConnectPortal->on(CONFIG_DONE_PAGE, std::bind<String>(&MyPortal::onAreaResults, this, std::placeholders::_1, std::placeholders::_2));
```

And I suspect it may be related to the fact that I return an "empty" string.

```c++
String MyPortal::onAreaResults(AutoConnectAux& aux, PageArgument& args){
    ...
    return String();
}
```

This is not impacting the behavior, but I wanted to make a note that this occurs and also try to understand what the callback is expected to return. `return String()` looks fishy. But, it is a common design pattern in most example files.

/Marcus

**Comment by Hieromon** [comment]
@faanskit 
> return String() looks fishy. But, it is a common design pattern in most example files.

The send member function of the WebServer class logs "`content length is zero`" because your AutoConnectAux is something wrong.
I have no idea if you are being dubious based on your legacy C knowledge, but if you mean that there are cases where RVO does not apply, [here](https://sigcpp.github.io/2020/06/08/return-value-optimization) is a way to verify it.
Please verify it yourself. If your MyPortal class code is properly coded, then `return String()` should be fine. It would be used rather aggressively.

And, Perhaps the answer to what you want to do is in `src/AutoConnectOTA.cpp` or `src/AutoConnectUpdate.cpp`. I recommend that you trace it thoroughly.

**Reply by faanskit** [reply]
When I access the page `.../area` from the sample code  you provided in https://github.com/Hieromon/AutoConnect/discussions/554#discussioncomment-4665271 I also get the _warning_:

It's one `send()` warning for each reload.

```
--- Terminal on COM4 | 115200 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, esp32_exception_decoder, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H

[ 14879][W][WebServer.cpp:436] send(): content length is zero
[ 15001][E][WebServer.cpp:649] _handleRequest(): request handler not found
[ 23168][W][WebServer.cpp:436] send(): content length is zero
[429298][W][WebServer.cpp:436] send(): content length is zero
[432367][W][WebServer.cpp:436] send(): content length is zero
[434208][W][WebServer.cpp:436] send(): content length is zero
```

Take note that I compile with `ARDUHAL_LOG_LEVEL_INFO` to see this warning:
```
[env:az-delivery-devkit-v4]
platform = espressif32
board = az-delivery-devkit-v4
framework = arduino
monitor_speed = 115200
upload_speed = 460800
lib_deps = hieromon/AutoConnect@^1.4.1
build_flags = 
	-DCORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_INFO
```


Even this basic page from [your documentation](https://hieromon.github.io/AutoConnect/basicusage.html)  throws the same warning:
```c++
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

static const char PAGE_HELLO[] = R"(
{
  "uri": "/",
  "title": "Hello",
  "menu": false,
  "element": [
    {
      "name": "caption",
      "type": "ACText",
      "value": "Hello, World"
    }
  ]
}
)";


WebServer Server;
AutoConnect Portal(Server);
void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();
  Portal.load(PAGE_HELLO);
  Portal.begin();
  Serial.println("Web Server started:" + WiFi.localIP().toString());
}

void loop() {
  Portal.handleClient();
}
```

```
Web Server started:192.168.1.193
[  5418][W][WebServer.cpp:436] send(): content length is zero
```


So, don't think it's my code that is causing this; even though it should be the first and usual suspect ;-)
And again, everything works as expected. Only this _warning_ that indicates something is not right.

/Marcus

**Reply by Hieromon** [reply]
?? Hm ??
Isn't that simply the start sign of a chunk? Because PageBuilder will chunk the transfer encoding if the content is not determined. It is not a problem with AutoConnect, rather the opposite, PageBuilder auto-chunking seems to work fine.

**Comment by Hieromon** [comment, marked as answer]
?? Hm ??
Isn't that simply the start sign of a chunk? Because PageBuilder will chunk the transfer encoding if the content length is not predetermined. It is not a problem with AutoConnect, nor of course with `return String()`. Rather, it is the other way around, and PageBuilder's auto-chanking seems to work fine.

**EDIT**

I have confirmed on the client browser side that the chunked transfer is successful.

**Reply by faanskit** [reply]
Great, I can determine it's not an issue with my code.

I guess this is a false positive from WebServer.cpp, which should not warn in this case. [RFC2616](https://greenbytes.de/tech/webdav/rfc2616.html#rfc.section.4.4) is even quite specific about it:
"_Messages must not include both a Content-Length header field and a non-identity transfer-coding. If the message does include a non-identity transfer-coding, the Content-Length must be ignored._"

Thanks for the clarification!

**Reply by Hieromon** [reply]
This has a history, and the ESP8266 WebServer class suppresses this warning; the ESP32 WebServer class has been modified to zero-check content length specifications, regardless of whether they are CONTENT_LENGTH_UNKNOWN or not.

https://github.com/espressif/arduino-esp32/blob/a95d838aab436588afb23656ef8190c3ff2e7b94/libraries/WebServer/src/WebServer.cpp#L430-L442

On the other hand, chunk transfer when dealing with the ESP8266WebServer class is:
https://github.com/Hieromon/PageBuilder/blob/6193e09a2eaafc71ace0a5651f18209f0cce2e7b/src/PageBuilder.cpp#L692-L696
```cpp
server.setContentLength(CONTENT_LENGTH_UNKNOWN);
server.send(code, "text/html", "");.
```

In fact, if the above blank transmission does not exist, the content transmission will not be complete.

However, sending empty content with Chunked Transfer Encoding essentially means a trailer. Let me quote from the RFC like you.

```
 Chunked-Body   = *chunk
                  last-chunk
                  trailer
                  CRLF

 chunk          = chunk-size [ chunk-extension ] CRLF
                  chunk-data CRLF
 chunk-size     = 1*HEX
 last-chunk     = 1*("0") [ chunk-extension ] CRLF

 chunk-extension= *( ";" chunk-ext-name [ "=" chunk-ext-val ] )
 chunk-ext-name = token
 chunk-ext-val  = token | quoted-string
 chunk-data     = chunk-size(OCTET)
 trailer        = *(entity-header CRLF)
```

I think there is a difference of opinion between ESP8266WebServer and ESP32WebServer as to which is the more appropriate implementation. (As for me, I don't care which one is better.) The low level library cannot determine if it is a redundant warning or not.
